spec/controllers/catalog_controller_spec.rb

```diff
@@ -0,0 +1,219 @@
+      @request.set_header "HTTP_X _REQUESTED_WITH", "XMLHttpRequest"
+      @request.set_header "HTTP_ACCEPT", "text/html"
+      expect {
+        get :previous_callnumber, params: { callnum: 'HB172.5%20.F692%202019', start: 1 }
+      }.to raise_error ActionController::UnknownFormat
+    end
+  end
+
+  context 'when both the Accept and X-Requested-With headers are missing' do
+    it 'raises an UnknownFormat error' do
+      expect {
+        get :previous_callnumber, params: { callnum: 'HB172.5%20.F692%202019', start: 1 }
+      }.to raise_error ActionController::UnknownFormat
+    end
+  end
+
+  context 'when headers and params are missing' do
+    it 'raises an ParameterMissing error' do
+      expect {
+        get :previous_callnumber
+      }.to raise_error ActionController::ParameterMissing
+    end
+  end
+
+  # /build_carousel
+  describe "GET carousel as xhr request and required params" do
+    it 'returns a 200' do
+      get :build_carousel, xhr: true, params: { callnum: 'HB172.5%20.F692%202019' }
+      expect(response).to have_http_status(:ok)
+    end
+  end
+
+  context 'when the callnum parameter is missing' do
+    it 'raises a ParameterMissing error' do
+      expect {
+        get :build_carousel, xhr: true
+      }.to raise_error ActionController::ParameterMissing
+    end
+  end
+
+  context 'when the X-Requested-With: XMLHttpRequest header is missing' do
+    it 'raises an InvalidCrossOriginRequest error' do
+      @request.set_header "HTTP_ACCEPT", "text/javascript"
+      expect {
+        get :build_carousel, params: { callnum: 'HB172.5%20.F692%202019' }
+      }.to raise_error ActionController::InvalidCrossOriginRequest
+    end
+  end
+
+  context 'when the Accept: text/javascript header is missing' do
+    it 'raises an InvalidCrossOriginRequest error' do
+      @request.set_header "HTTP_X _REQUESTED_WITH", "XMLHttpRequest"
+      expect {
+        get :build_carousel, params: { callnum: 'HB172.5%20.F692%202019' }
+      }.to raise_error ActionController::UnknownFormat
+    end
+  end
+
+  context 'when the Accept header does not include javascript' do
+    it 'raises an InvalidCrossOriginRequest error' do
+      @request.set_header "HTTP_X _REQUESTED_WITH", "XMLHttpRequest"
+      @request.set_header "HTTP_ACCEPT", "text/html"
+      expect {
+        get :build_carousel, params: { callnum: 'HB172.5%20.F692%202019' }
+      }.to raise_error ActionController::UnknownFormat
+    end
+  end
+
+  context 'when both the Accept and X-Requested-With headers are missing' do
+    it 'raises an UnknownFormat error' do
+      expect {
+        get :build_carousel, params: { callnum: 'HB172.5%20.F692%202019' }
+      }.to raise_error ActionController::UnknownFormat
+    end
+  end
+
+  context 'when headers and params are missing' do
+    it 'raises an ParameterMissing error' do
+      expect {
+        get :build_carousel
+      }.to raise_error ActionController::ParameterMissing
+    end
+  end
+
+end
```

Comment:
Nice job getting these working controller specs - it sounds like you did some research that suggested request specs might be the recommended way now, was that something you experimented with as well? I saw that the rspec documentation actually recommends creating request specs when testing headers (bottom of https://rspec.info/features/6-0/rspec-rails/controller-specs/). But I think having these controller specs is already a big improvement.

Reply:
yes, i tried request specs but it gets complicated when the api is an external one
once i get a bit more experience i'll try again but as you say, this is a slight improvement